frontend, libobs, plugins: Natural sorting for files, profile and scene names

Dankirk opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
Applies natural, locale aware sorting when drag&dropping many sources at once and to profile and scene names in menus. For slideshow (mk2) and VLC the files in added directories are cycled using this sorting method instead of an arbitrary order. The order of items in the slideshow listbox is otherwise preserved.

For background: Natural sort means sorting by numerical value. Alphabetical sort arranges numbers like this: 1 < 10 < 11 < 2 < 3.
Natural sort does it this way: 1 < 2 < 3 < 10 < 11.
The natural sort is what Windows explorer uses under the hood when you sort by name or use `ls -v` on unix.

For general use, `qt-wrappers` has couple new methods for natural sorting:
```
template<typename T, typename Callback> void NaturalSort(QList<T> &list, Callback extractor, bool reverse = false)
template<typename T> void NaturalSort(QList<T> &list, bool reverse = false)
```
The list parameter is sorted using locale and number value aware algorithm with `QCollator`. It works with `QString` and `QStringView`. System locale is used to achieve the locale-aware sorting.

For lower level stuff in C, like slideshow plugin, there's new comparison functions for `char*` and `wchar_t*` in `dstr` library.
```
EXPORT int astrnatcmp(const char *str1, const char *str2, os_locale_t locale);
EXPORT int wstrnatcmp(const wchar_t *str1, const wchar_t *str2, os_locale_t locale);
```
The return value is <0, 0 or >0 as expected for return value of `qsort()`'s callback function. This sorting uses `strcoll()` and `wcscoll()` respectively for locale aware sorting, but does natural number sorting using a custom implementation.

OBS doesn't currently use process wide system locale for C/C++ on Windows (issue disussed here #12624), so new locale apis are added to libobs platform library and documentation added. These can be used to do locale aware comparison / sorting in C directly and are used as basis for the main things `astrnatcmp` and `wstrnatcmp`.

```
EXPORT int os_strcoll_l(const char *a, const char *b, os_locale_t locale);
EXPORT int os_wcscoll_l(const wchar_t *a, const wchar_t *b, os_locale_t locale);
EXPORT os_locale_t os_get_locale(const char *locale);
EXPORT void os_free_locale(os_locale_t locale);
EXPORT void os_locale_aware_sort(void *base, size_t elements, size_t element_size, os_locale_aware_cmp cmp,
				 os_locale_t locale);
```

_OBSBasic_Dropfiles.cpp_ has somewhat sizable changes, due to how late the names for listed sources were acquired. This changes it so names are gathered first for sorting and only later added to the UI.

### Motivation and Context
Fixes #10382 
Couple requests on forums [here](https://obsproject.com/forum/threads/image-slideshow-plays-in-seemingly-random-order-google-drive-issue.152786/#post-680967) and [here](https://obsproject.com/forum/threads/image-slide-show-does-not-display-exported-powerpoint-slides-in-order-using-add-directory.184255/). Most recently I saw a [Youtuber mention](https://youtu.be/LMXky-XlUqM?t=2252) this during a live stream.

The issue is with filenames and other listings that contain incremental numbers which are not sorted intuitively when sorted alphabetically. In fact the alphabetical sorting is also coincidental, since the methods for iterating over files give no guarantee about the order.

### How Has This Been Tested?
I tested the features with files in directories with the following file names to see if they are sorted correctly:
[testnames.txt](https://github.com/user-attachments/files/22163540/testnames.txt)
The files should appear in the order listed for en_FI locale. The files with 'ä' or 'ö' could be in another place for different locales. The important part is the numerical comparison works, so _5.jpg_ is before _10.jpg_ while the opposite is true for basic alphabetical sorting.

I'm on Windows 11 English US version, but with Finnish region/locale settings.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Dankirk commented:
Updated the list of filenames I tested the sorting with and I'm happy with the results.

One issue remains:
For image slideshow directories, the sorting is not locale aware (natural sort works), because the app uses the default 'C' locale. The app does call QLocale::setDefault() for Qt based things during app start, but does not call setlocale() for lower level functions. For sorting purposes setting LC_COLLATE with setlocale() would suffice, but it is uncertain how much that affects other things. Parts that can use OBSBasic class, use QCollator instead with system locale.

There is also a question if the default locale set with QLocale should actually be the system locale instead of the app language/translations locale? If so, could we then also do so with setlocale() to match?

Dankirk commented on frontend/widgets/OBSBasic.hpp:
```diff
@@ -457,14 +460,39 @@ private slots:
 
+public:
+	template<typename T, typename Callback>
+	static void naturalSort(QList<T> &list, Callback extractor, bool reverse = false)
```
Moved to qt-wrappers header

Dankirk commented on libobs/util/dstr.h:
```diff
@@ -51,6 +51,12 @@ EXPORT int astrcmp_n(const char *str1, const char *str2, size_t n);
+static inline int astrncoll(const char *str1, const char *str2, size_t read);
+static inline int wstrncoll(const wchar_t *str1, const wchar_t *str2, size_t read);
+static inline int a_compare_number(const char *a, const char *b, size_t *j);
+static inline int w_compare_number(const wchar_t *a, const wchar_t *b, size_t *j);
```
Moved to dstr.c

norihiro commented on docs/sphinx/reference-libobs-util-dstr.rst:
```diff
@@ -59,6 +59,28 @@ General String Helper Functions
 
 ----------------------
 
+.. function:: int astrnatcmp(const char *str1, const char *str2)
+
+   Natural comparison function for strings.
+
+   In natural comparison numbers are compared by their value instead of the alphabetical order.
+   For example "z2" < "z11" while in alphabetical comparison it would be "z11" < "z2".
+
+   Otherwise the function uses locale aware string comparison function strcol().
+
+----------------------
+
+.. function:: int wstrnatcmp(const wchar_t *str1, const wchar_t *str2)
+
+   Natural comparison function for wide strings.
+
+   In natural comparison numbers are compared by their value instead of the alphabetical order.
+   For example "z2" < "z11" while in alphabetical comparison it would be "z11" < "z2".
+
+   Otherwise the function uses locale aware wide string comparison function wcscoll().
+
+----------------------
+
```
`astrnatcmp("01", "1")` returns -1. However, `strnatcmp("01", "1")` in PHP returns 0, for example.
Is this your intention?

Dankirk replied:
Yes, this is intentional. `a_compare_number()` returns 0 for those inputs, so `astrnatcmp()` will use `strcol()` to break the tie (until next number pair is encountered in the strings if any). This results in same ordering as Windows uses in Explorer when ordering by name or `ls -v` for linux

norihiro commented on libobs/util/dstr.c:
```diff
@@ -163,6 +170,239 @@ int wstrcmpi_n(const wchar_t *str1, const wchar_t *str2, size_t n)
+	return 4;
+}
+
+// Windwos wchar_t is 16bit while unix is 32bit, thus Windows needs to check for surrogates
```
```suggestion
// Windows wchar_t is 16bit while unix is 32bit, thus Windows needs to check for surrogates
```

Dankirk committed:
frontend: Natural sort drag&drop sources

Sort drag&dropped sources using locale-aware natural sorting with QCollator and system locale.

Dankirk committed:
frontend: Natural sort profiles and scenes listing in menus

Change locale-aware sorting of profiles and scene collections in menus to locale-aware natural sorting, so number sequences in names are treated by value rather than their alphabetical order.

Dankirk committed:
libobs: Add natural comparison methods for strings

New comparison methods astrnatcmpi() and wstrnatcmpi() for sorting strings and documentation for them. Uses locale aware strcol() / wcscoll() and a custom implementation of natural comparison for number sequences. Supports multibyte and whcar surrogate characters.

This is intended to be used for parts that do not have direct access to higher level methods like Qt QCollator.

Dankirk committed:
plugins: Natural sort slideshow mk2 images in directories

Applies locale-aware natural sorting to the cycled files when adding directory to image slideshow. Previously files in directories were cycled in arbitrary order by os_readdir().

Locale-awareness is limited on Windows currently, because we are using minimal 'C' locale for it. This is to be addressed separately.

Warchamp7 reviewed (approved):
Approving on a conceptual/UX level, but others will need to speak to the technical implications

Dankirk committed:
vlc-video: Apply natsort to vlc playlist dirs

Applies the same natural, alphabetical sort to vlc playlist directories that was added for slideshow.

Dankirk committed:
libobs, plugins: Add locale functions for libobs

Adds locale aware functions for libobs and makes slideshow and vlc plugins use them for natural, locale aware file sorting.

Adds documentation for the new locale apis.
